src/dotnet/Notification.Service/Notifications.Backend.cs

```diff
@@ -111,26 +117,45 @@ public virtual async Task<ImmutableArray<Symbol>> ListSubscriberIds(string chatI
             .SelectMany(uid => GetDevicesInternal(uid, cancellationToken))
             .Chunk(200, cancellationToken);
 
+        var usersToInvalidate = new HashSet<string>();
         await foreach (var deviceGroup in deviceIdGroups.ConfigureAwait(false)) {
             multicastMessage.Tokens = deviceGroup.Select(p => p.DeviceId).ToList();
             var batchResponse = await _firebaseMessaging.SendMulticastAsync(multicastMessage, cancellationToken)
                 .ConfigureAwait(false);
 
             if (batchResponse.FailureCount > 0) {
-                var errorInfoItems = await Task.WhenAll(batchResponse.Responses
-                    .Where(r => !r.IsSuccess)
-                    .Select(async r => new {
-                        r.Exception.MessagingErrorCode,
-                        r.Exception.HttpResponse.StatusCode,
-                        Content = await r.Exception.HttpResponse.Content.ReadAsStringAsync(cancellationToken)
-                            .ConfigureAwait(false),
-                    })).ConfigureAwait(false);
-                var errorDetails = errorInfoItems
-                    .Select(x => $"MessagingError = {x.MessagingErrorCode}; HttpCode = {x.StatusCode}; Content = {x.Content}")
-                    .Aggregate(new StringBuilder(), (sb, item) => sb.AppendLine(item))
-                    .ToString();
-                _log.LogWarning("Notification messages were not sent. NotificationCount = {NotificationCount}; {Details}",
-                    batchResponse.FailureCount, errorDetails);
+                var responses = batchResponse.Responses
+                    .Zip(deviceGroup)
+                    .Select(p => new {
+                        p.Second.DeviceId,
+                        p.Second.UserId,
+                        p.First.IsSuccess,
+                        p.First.Exception?.MessagingErrorCode,
+                        p.First.Exception?.HttpResponse,
+                    })
+                    .ToList();
+                var responseGroups = responses
+                    .GroupBy(x => x.MessagingErrorCode);
+                foreach (var responseGroup in responseGroups)
+                    if (responseGroup.Key == MessagingErrorCode.Unregistered) {
+                        var tokensToRemove = responseGroup
```

Comment:
It's a bit strange that NotifySubscribers method invalidates ListDevices result.
I suppose it make sense to refactor it in the following way:
1) introduce command RemoveDevicesCommand(ImmutableArray<string> deviceIds)
2) during processing response results collect unregistered device ids.
3) after all responses are processed, create RemoveDevicesCommand and invoke it.

It should be responsibility of RemoveDevicesCommand handler to invalidate ListDevices result.
Also I see possible issue with executing in background RemoveOutdatedTokens method. In fact we can have a race condition situation that ListDevices result invalidation may happen before RemoveOutdatedTokens actually executed. It should be solved eventually but still approach with RemoveDevicesCommand handler looks more predictable to me.